packages/workflow/index.ts

```diff
@@ -0,0 +1,159 @@
+}
+
+export class UpscalerTask extends WorkflowTask {
+  private image_url: string;
+
+  constructor(options: UpscalerTaskOptions) {
+    super(options);
+    this.image_url = options.image_url;
+  }
+
+  get task_type(): WorkflowTaskType {
+    return WorkflowTaskType.Upscaler;
+  }
+
+  get task_details(): Record<string, any> {
+    return { parameters: { image: this.image_url } };
+  }
+}
+
+export interface WorkflowTaskResult {
+  task_id: string
+  status: 'waiting' | 'running' | 'finished' | 'failed'
+  result?: any
+}
+
+export class Workflow extends APIResource {
+  async executeWorkflow(task: WorkflowTask): Promise<string> {
+    await this.resourceRequest(task.consumer_id)
+    const task_id = await this.createTask(task)
+    return task_id;
+  }
+
+  async queryTaskResult(task_id: string): Promise<WorkflowTaskResult> {
+    const url = `${this._client.workflowURL}/task_result_query`
+    const headers = {
+      'Content-Type': 'application/json',
+      Authorization: `Bearer ${this._client.apiKey}`,
+    }
+    const data = { task_id }
+
+    const response = await fetch(url, {
+      method: 'POST',
+      headers,
+      body: JSON.stringify(data),
+    })
+
+    if (!response.ok) {
+      throw new Error(`Task result query failed: ${response.statusText}`)
+    }
+
+    return await response.json()
+  }
+
+  async resourceRequest(consumer_id: string): Promise<string> {
+    const url = `${this._client.workflowURL}/resource_request`
+    const headers = {
+      'Content-Type': 'application/json',
+      Authorization: `Bearer ${this._client.apiKey}`,
+    }
+    const data = { consumer_id }
+
+    const response = await fetch(url, {
+      method: 'POST',
+      headers,
+      body: JSON.stringify(data),
+    })
+
+    if (!response.ok) {
+      throw new Error(`Resource request failed: ${response.statusText}`)
+    }
+
+    const result = await response.json()
+    return result.miner_id
+  }
+
+  async createTask(task: WorkflowTask): Promise<string> {
+    const url = `${this._client.workflowURL}/task_create`
+    const headers = {
+      'Content-Type': 'application/json',
+      Authorization: `Bearer ${this._client.apiKey}`,
+    }
+    const { job_id_prefix = 'sdk-workflow' } = task;
+    const id = Randomstring.generate({ charset: 'hex', length: 10 })
+    const job_id = `${job_id_prefix}-${id}`
+
+    const data = {
+      consumer_id: task.consumer_id,
+      task_type: task.task_type,
+      task_details: task.task_details,
+      workflow_id: task.workflow_id,
+      job_id
+    };
+
+    const response = await fetch(url, {
+      method: 'POST',
+      headers,
+      body: JSON.stringify(data),
+    })
+
+    if (!response.ok) {
+      throw new Error(`Task creation failed: ${response.statusText}`)
+    }
+
+    const result = await response.json()
+    return result.task_id
+  }
+
+  async executeWorkflowAndWaitForResult(
+    task: WorkflowTask,
+    timeout: number = 300000,
+    interval: number = 10000
```

Comment:
Let's set a const MIN_INTERVAL = 1000 ms to restrict the interval being too short and may effectively create a DDOS